src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AttachedCollections/NuGet/Models/AssetsFileTarget.cs

```diff
@@ -0,0 +1,183 @@
+{
+    /// <summary>
+    /// Immutable snapshot of data captured from <c>project.assets.json</c> that relates to a specific target.
+    /// </summary>
+    internal sealed class AssetsFileTarget
+    {
+        /// <summary>
+        /// Gets the target name, such as <c>.NETFramework,Version=v4.8</c> or <c>.NETStandard,Version=v1.3</c>.
+        /// </summary>
+        public string Target { get; }
+
+        /// <summary>
+        /// Gets diagnostic messages for this target. Often empty.
+        /// </summary>
+        public ImmutableArray<AssetsFileLogMessage> Logs { get; }
+
+        /// <summary>
+        /// Stores data about libraries (packages/projects), keyed by name.
+        /// </summary>
+        public ImmutableDictionary<string, AssetsFileTargetLibrary> LibraryByName { get; }
+
+        /// <summary>
+        /// The snapshot to which this target data belongs.
+        /// </summary>
+        private readonly AssetsFileDependenciesSnapshot _snapshot;
+
+        /// <summary>
+        /// Lazily populated cache of back-references in the dependencies graph, mapping libraries to their ancestor(s).
+        /// Created by <see cref="TryGetDependents"/>, which is used during computation of Solution Explorer search results.
+        /// </summary>
+        private IReadOnlyDictionary<string, ImmutableArray<AssetsFileTargetLibrary>>? _dependentsByLibrary;
+
+        /// <summary>
+        /// Lazily populated cache of dependencies as <see cref="AssetsFileTargetLibrary"/> objects.
+        /// </summary>
+        private readonly Dictionary<(string LibraryName, string? Version), ImmutableArray<AssetsFileTargetLibrary>> _dependenciesByNameAndVersion = new Dictionary<(string LibraryName, string? Version), ImmutableArray<AssetsFileTargetLibrary>>();
+
+        public AssetsFileTarget(AssetsFileDependenciesSnapshot snapshot, string target, ImmutableArray<AssetsFileLogMessage> logs, ImmutableDictionary<string, AssetsFileTargetLibrary> libraryByName)
+        {
+            Requires.NotNull(snapshot, nameof(snapshot));
+            Requires.NotNullOrWhiteSpace(target, nameof(target));
+            Requires.Argument(!logs.IsDefault, nameof(logs), "Must not be default");
+            Requires.NotNull(libraryByName, nameof(libraryByName));
+
+            Target = target;
+            _snapshot = snapshot;
+            Logs = logs;
+            LibraryByName = libraryByName;
+        }
+
+        /// <summary>
+        /// Gets the set of dependents (parents) of <paramref name="libraryName"/>.
+        /// </summary>
+        /// <returns><see langword="true"/> if dependents were found, otherwise <see langword="false"/>.</returns>
+        public bool TryGetDependents(string libraryName, out ImmutableArray<AssetsFileTargetLibrary> dependents)
+        {
+            // Defer construction of dependents collection until needed. It's only needed for Solution Explorer search.
+            if (_dependentsByLibrary == null)
```

Comment:
Can this overlap?

Reply:
Only if operations performed within `Enumerable.ToDictionary` are reordered after the field write, which seems unlikely given the stack pop. Using `Volatile.Write` to the field just in case.